### PR TITLE
Add SeletMeny and SelectMenyBy

### PR DESCRIPTION
**Pull request by mkitawaki**
### Flattens the resulting sequences into one sequence

- SelectMeny(f func(in T,idx int) (T, error)) (r Query)

  - in:element of original source
  - idx: index of element

- SelectMenyBy(f func(in T,idx int) (T, error), resultSelector func(parent T, child T) (T, error)) (r Query)

  - in: element of original source
  - idx: index of element
  - parent: element of origin source
  - child: result of invoking the transform function on each element of original source

### CODE REVIEW

**Comment by ahmetalpbalkan** [comment]
Hi @mkitawaki, thanks for the change. I will get back to reviewing this very soon. Sorry for the delay. Meanwhile can you please squash 2 commits into 1 and push -f to your branch to update the PR?

File: linq.go [hunk 1]
```diff
@@ -1291,3 +1291,57 @@ func (q Query) MaxFloat64() (min float64, err error) {
 	}
 	return q.values[maxIndex].(float64), nil
 }
+
+// SelectMany returns flattens the resulting sequences into one sequence.
+//
+// Example:
+// 	names, err := From(parents).SelectMany(func (p T,idx int) (T, error){
```
Comment (ahmetalpbalkan) [inline linq.go]: space missing before comma `, idx`

File: linq.go [hunk 2]
```diff
@@ -1291,3 +1291,57 @@ func (q Query) MaxFloat64() (min float64, err error) {
...
+	})
+}
+
+// SelectMany returns flattens the resulting sequences into one sequence.
+//
+// resultSelector takes parent element and child element as inputs
+// and returns a value which will be an element in the resulting query.
+//
+// Example:
+// 	names, err := From(parents).SelectManyBy(func (p T,idx int) (T, error){
```
Comment (ahmetalpbalkan) [inline linq.go]: space missing before comma: `, idx`

File: linq.go [hunk 3]
```diff
@@ -1291,3 +1291,57 @@ func (q Query) MaxFloat64() (min float64, err error) {
...
+//		return p.(*Parent).Children, nil
+// 	},func (p T, c T) (T, error){
+//              return p.(*Parent).Name + ":" + c.(*Child).Name
```
Comment (ahmetalpbalkan) [inline linq.go]: should this end with a `, nil`? it seems like it returns `(T, error)`

File: linq.go [hunk 4]
```diff
@@ -1291,3 +1291,57 @@ func (q Query) MaxFloat64() (min float64, err error) {
...
+
+// SelectMany returns flattens the resulting sequences into one sequence.
+//
+// resultSelector takes parent element and child element as inputs
+// and returns a value which will be an element in the resulting query.
+//
+// Example:
+// 	names, err := From(parents).SelectManyBy(func (p T,idx int) (T, error){
+//		return p.(*Parent).Children, nil
+// 	},func (p T, c T) (T, error){
```
Comment (ahmetalpbalkan) [inline linq.go]: space before `{`

**Comment by mkitawaki** [comment]
Hi @ahmetalpbalkan, thanks for comment.I updated a source file.
Please kindly confirm it.

File: linq.go [hunk 5]
```diff
@@ -1291,3 +1291,56 @@ func (q Query) MaxFloat64() (min float64, err error) {
...
+		return
+	}
+
+	for i, p := range q.values {
+		val, err := f(p, i)
+		if err != nil {
+			r.err = err
+			return r
+		}
+		for _, c := range From(val).values {
```
Comment (ahmetalpbalkan) [inline linq.go]: usage of `From(...).values` seems unnecessary here Why not just `range val`?

File: linq.go [hunk 6]
```diff
@@ -1291,3 +1291,56 @@ func (q Query) MaxFloat64() (min float64, err error) {
...
 	return q.values[maxIndex].(float64), nil
 }
+
+// SelectMany returns flattens the resulting sequences into one sequence.
+//
+// Example:
+// 	names, err := From(parents).SelectMany(func (p T, idx int) (T, error) {
+//		return p.(*Parent).Children, nil
+// 	}).Results()
+func (q Query) SelectMany(f func(T, int) (T, error)) (r Query) {
```
Comment (ahmetalpbalkan) [inline linq.go]: can you move these methods to just after `Select()`? It logically makes sense to have them next to each other. (same with plinq.go)

File: linq_test.go [hunk 7]
```diff
@@ -1329,3 +1329,135 @@ func TestMinMax(t *testing.T) {
 		})
 	})
 }
+
+type bar struct {
+        str  string
+        foos []foo
+}
+type fooBar struct {
+        fooStr string
```
Comment (ahmetalpbalkan) [inline linq_test.go]: this line is not properly gofmt 'ed

File: linq_test.go [hunk 8]
```diff
@@ -1329,3 +1329,135 @@ func TestMinMax(t *testing.T) {
...
+	barStr string
```
Comment (ahmetalpbalkan) [inline linq_test.go]: this line is not properly gofmt 'ed. please run gofmt on all code.

File: linq_test.go [hunk 9]
```diff
@@ -1329,3 +1329,135 @@ func TestMinMax(t *testing.T) {
...
+}
+
+var     (
+	fooArr  = []foo{foo{"A", 0}, foo{"B", 1}, foo{"C", -1}}
```
Comment (ahmetalpbalkan) [inline linq_test.go]: again not properly gofmt'ed

File: linq.go (hunk 5 above)
Comment (mkitawaki) [inline linq.go]: val is not []T.
Should I change from `From(val).values` to `result of takeSliceArg`?

Comment (ahmetalpbalkan) [inline linq.go]: @mkitawaki hmm I don't exactly remember but `T` is just an alias for `interface{}` so anything automatically can be converted to `T`.

In this case I think you're trying to build a `[]T`, in that case I think you're looking for `takeSliceArg()`.

Comment (mkitawaki) [inline linq.go]: I changed from `From(val).values` to `takeSliceArg()`.

File: linq.go (hunk 6 above)
Comment (mkitawaki) [inline linq.go]: I have moved these methods to just after Select().

File: linq_test.go (hunk 8 above)
Comment (mkitawaki) [inline linq_test.go]: I ran gofmt on all code.

File: linq_test.go (hunk 9 above)
Comment (mkitawaki) [inline linq_test.go]: I ran gofmt on all code.

File: linq_test.go [hunk 10]
```diff
@@ -226,6 +226,138 @@ func TestSelect(t *testing.T) {
...
+	str  string
+	foos []foo
+}
+type fooBar struct {
+	fooStr string
+	barStr string
+}
+
+var (
+	fooArr    = []foo{foo{"A", 0}, foo{"B", 1}, foo{"C", -1}}
```
Comment (ahmetalpbalkan) [inline linq_test.go]: You can remove the "foo" keyword of struct if you define the slice as []foo.

Comment (ahmetalpbalkan) [inline linq_test.go]: But we can keep it, it's not harmful.

**Comment by ahmetalpbalkan** [comment]
Thanks @mkitawaki  I will tag this change with a new version soon.

**Pull request merged**